Spark date_add and date_sub to binary operators

Katakam-Phani-Mahidhar opened the pull request.

deepakkadam-dm reviewed (commented):
1. Latest master merge needed in this branch

2. Function name BINARY_DATE_ADD/BINARY_DATE_SUB is misleading. 
It falsely indicates that the operands for the function are Date(s). And the function will add/subtract the dates.
E.g. date1 + date2
Whereas in reality the function adds DAYS to the input Date.
inputDate + 2 => inputDate + 2 DAYS

3. The unparse() logic in other Functions/Operators is delegated to the Dialect. In case of these two functions it is implemented in the SqlLibraryOperators.

4. The logic in SparkSqlDialect.getTargetFunc() can be changed to follow the one in BigQuerySqlDialect
i.e. First check for Call's Operator then rest (Operand type, etc) of the things. Looks more logical

Katakam-Phani-Mahidhar commented:
> 1. Latest master merge needed in this branch
> 2. Function name BINARY_DATE_ADD/BINARY_DATE_SUB is misleading.
>    It falsely indicates that the operands for the function are Date(s). And the function will add/subtract the dates.
>    E.g. date1 + date2
>    Whereas in reality the function adds DAYS to the input Date.
>    inputDate + 2 => inputDate + 2 DAYS
> 3. The unparse() logic in other Functions/Operators is delegated to the Dialect. In case of these two functions it is implemented in the SqlLibraryOperators.
> 4. The logic in SparkSqlDialect.getTargetFunc() can be changed to follow the one in BigQuerySqlDialect
>    i.e. First check for Call's Operator then rest (Operand type, etc) of the things. Looks more logical

1. rebase the branch.
2. changed the operators name from binary_date_add, binary_date_sub to date_add_day, date_sub_day.
3. It has been suggested that the unparsing logic of individual dialect specific operators should be contained in themselves. Please do tell if the unparse logic of the operators should be present in SparkSqlDialect instead.
4. Updated targetFunc to follow the same format as of BigQuery. i.e., operator kind check first and only then the remaining checks as suggested.

soma-mondal commented:
Hi, There is already functions SqlLibraryOperators.DATE_SUB and SqlLibraryOperators.DATE_ADD (which take care of interval operations). Please use already present functions.

Katakam-Phani-Mahidhar commented:
> Hi, There is already functions SqlLibraryOperators.DATE_SUB and SqlLibraryOperators.DATE_ADD (which take care of interval operations). Please use already present functions.

Hi. DATE_ADD and DATE_SUB operators translates to DATE_ADD and DATE_SUB functions respectively, but UAT team wants it as a binary oeprator. 
i.e., source:  SELECT CURRENT_DATE + 1;
       expected Target: SELECT CURRENT_DATE + 1;

soma-mondal commented:
@Katakam-Phani-Mahidhar Please override unparseIntervalOperandsBasedFunctions method to achieve the same

Katakam-Phani-Mahidhar commented:
> @Katakam-Phani-Mahidhar Please override unparseIntervalOperandsBasedFunctions method to achieve the same

done

soma-mondal commented on core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:
```diff
@@ -371,6 +385,12 @@ public SparkSqlDialect(SqlDialect.Context context) {
     case LITERAL:
     case TIMES:
+      String opName = call.getOperator().toString();
+      if (opName.equalsIgnoreCase("DATE_SUB")
```
Please add a case

Katakam-Phani-Mahidhar replied:
done

soma-mondal commented on core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:
```diff
@@ -408,7 +440,7 @@ private void unparseIntervalOperandCall(
    */
 
-  private SqlNode modifySqlNode(SqlWriter writer, SqlNode intervalOperand) {
+  private static SqlNode modifySqlNode(SqlWriter writer, SqlNode intervalOperand) {
```
Please remove static checks from methods

Katakam-Phani-Mahidhar replied:
done

soma-mondal commented on core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:
```diff
@@ -389,6 +409,18 @@ private void unparseIntervalOperandCall(
+      writer.sep("+");
+    }
+    SqlNode intervalValue = modifySqlNode(writer, call.operand(1));
+    writer.print(intervalValue.toString().replace("`", ""));
```
Please use unparse method for intervalValue

Katakam-Phani-Mahidhar replied:
done

soma-mondal commented on core/src/main/java/org/apache/calcite/util/interval/SparkDateTimestampInterval.java:
```diff
@@ -67,12 +67,13 @@ public boolean unparseDateTimeMinus(SqlWriter writer, SqlCall call,
       case "INTERVAL_MONTH":
         handleIntervalMonth(writer, call, leftPrec, rightPrec, sign);
         break;
-      case "INTERVAL_DAY":
```
Why this change is required ?

Katakam-Phani-Mahidhar replied:
This was actually overriding the unparse method of DATE_ADD, DATE_SUB operators so it was initially removed, but since TIMESTAMP_ADD, TIMESTAMP_SUB operators are coming to this switch statement so added it back and written DATE_ADD, DATE_SUB specific logic instead.

soma-mondal commented on core/src/main/java/org/apache/calcite/util/interval/SparkDateTimestampInterval.java:
```diff
@@ -67,12 +67,13 @@ public boolean unparseDateTimeMinus(SqlWriter writer, SqlCall call,
       case "INTERVAL_SECOND":
         handleIntervalDatetimeUnit(writer, call, leftPrec, rightPrec, sign);
         break;
+      default:
```
this one as well

Katakam-Phani-Mahidhar replied:
removed it.
